controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,219 @@
+package namespace
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/red-hat-storage/ocs-operator/controllers/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// once is a pointer to the sync.Once type
+var once sync.Once
+
+// NamespaceReconciler reconciles the namespaces starting with openshift-*
+// nolint:revive
+type NamespaceReconciler struct {
+	client.Client
+	ctx    context.Context
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=list;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch;update;patch
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// i.e,the annotations of the namespace
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	prevLogger := r.Log
+	defer func() { r.Log = prevLogger }()
+	r.Log = r.Log.WithValues("Request.Name", request.Name)
+	r.ctx = ctx
+
+	r.Log.Info("Reconciling Namespace.")
+
+	var err error
+	updated := false
+
+	once.Do(func() {
+		updated, err = r.annotateOpenshiftNamespaces()
+		if err != nil {
+			return
+		}
+		if updated {
+			err = r.restartCSIAddonsPod()
+		}
+	})
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// If it's not the first reconcile and the namespaces were not updated in the once.Do func call
+	if !updated {
+		ns := &corev1.Namespace{}
+		err = r.Client.Get(ctx, request.NamespacedName, ns)
+		if err != nil {
+			// the namespace is probably deleted
+			if errors.IsNotFound(err) {
+				r.Log.Error(err, fmt.Sprintf("Namespace %s does not exist", request.Name))
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+
+		// Annotate namespaces starting with openshift-*, but ignore namespace if
+		// it has annotation "'skipReclaimspaceSchedule':'true'" set
+		updated, err := r.annotateNamespaces([]corev1.Namespace{*ns})
+		if err != nil {
+			r.Log.Error(err, fmt.Sprintf("Failed to annotate namespace: %s", ns.Name))
+			return reconcile.Result{}, err
+		}
+
+		// Restart the csi-addons controller pod if any namespace was annotated
+		if updated {
+			err = r.restartCSIAddonsPod()
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
```

Comment:
if you move this outside, you can have one instance of this code to cover both cases 1. In case of Once.Do and for single namespace Reconcile.

Reply:
Ahh, right. Updated the code accordingly.